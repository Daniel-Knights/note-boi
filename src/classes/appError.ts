export const ERROR_CODE = {
  NONE: 0,
  UNKNOWN: 1,
  LOGIN: 2,
  SIGNUP: 3,
  LOGOUT: 4,
  // PUSH was 5
  // PULL was 6
  DELETE_ACCOUNT: 7,
  CHANGE_PASSWORD: 8,
  ENCRYPTOR: 9,
  FORM_VALIDATION: 10,
  SYNC: 11,
} as const;

type ErrorCodes = typeof ERROR_CODE;
type ErrorKey = keyof ErrorCodes;
type ErrorCode = ErrorCodes[ErrorKey];

// Using `unknown` here breaks inference
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ErrorConfig<T extends (...args: any) => void> = {
  code: ErrorCode;
  message?: string;
  originalError?: unknown;
  /** Function to call on 'Try again', and its arguments */
  retry?: {
    fn: T;
    args?: Parameters<T>;
  };
  /** Whether to display the error in a form and/or `SyncStatus` */
  display?: {
    form?: boolean;
    sync?: boolean;
  };
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export class AppError<T extends (...args: any) => void = (...args: any) => void> {
  readonly code;
  readonly message;
  readonly display;
  readonly retryConfig;
  readonly originalError;

  constructor(config: ErrorConfig<T> = { code: ERROR_CODE.NONE }) {
    this.code = config.code;
    this.message = config.message;
    this.display = Object.freeze(config.display);
    this.retryConfig = Object.freeze(config.retry);
    this.originalError = config.originalError;

    this.retry = this.retry.bind(this);
  }

  get isNone() {
    return this.code === ERROR_CODE.NONE;
  }

  retry() {
    if (!this.retryConfig) return;

    const { fn, args } = this.retryConfig;

    return args ? fn(args) : fn();
  }
}
