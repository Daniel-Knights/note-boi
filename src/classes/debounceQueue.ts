/**
 * Queues and debounces async function calls.
 * If a function is cancelled after it has already begun executing, `isCancelled` can
 * be used to check if it should stop.
 */
export class DebounceQueue {
  #cancelled = new Map<number, { isRunning: boolean }>();
  #current: { id: number | null; isRunning: boolean } = {
    id: null,
    isRunning: false,
  };

  /**
   * Adds new task to the debounce queue.
   * If there is an existing task, it will be cleared.
   */
  add(cb: (isCancelled: () => boolean) => Promise<void>, delay?: number) {
    this.clear();

    const timeoutId = window.setTimeout(() => {
      if (timeoutId !== this.#current.id) return;

      this.#current.isRunning = true;

      cb(() => this.#isCancelled(timeoutId)).finally(() => {
        this.#cancelled.set(timeoutId, { isRunning: false });
      });
    }, delay);

    this.#current.id = timeoutId;
    this.#current.isRunning = false;
  }

  /**
   * Clears the current task in the debounce queue.
   * If the task is running, it will be marked as cancelled.
   */
  clear() {
    if (!this.#current.id) return;

    if (this.#current.isRunning) {
      this.#cancelled.set(this.#current.id, { isRunning: true });
    } else {
      window.clearTimeout(this.#current.id);

      this.#current.id = null;
    }
  }

  /**
   * Checks if task with the given timeout ID has been cancelled.
   */
  #isCancelled(timeoutId: number) {
    const cancelled = !!this.#cancelled.get(timeoutId);

    this.#cancelled.delete(timeoutId);

    return cancelled;
  }
}
