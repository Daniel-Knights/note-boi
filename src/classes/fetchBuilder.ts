import { Endpoint, EndpointPayloads } from '../constant';
import { ParsedResponse } from '../store/sync';
import { isDev, tauriInvoke } from '../utils';

export class FetchBuilder<
  E extends Endpoint = Endpoint,
  R = EndpointPayloads[E]['response'],
> {
  #endpoint;
  #init: RequestInit = {};

  constructor(e: E) {
    this.#endpoint = e;
  }

  static serverUrl =
    isDev() || process.env.NODE_ENV === 'test'
      ? 'http://localhost:8000'
      : 'https://note-boi-server-1098279308841.europe-west2.run.app';

  static defaultHeaders = {
    'Content-Type': 'application/json',
    'Content-Security-Policy': `default-src 'self'; connect-src ${FetchBuilder.serverUrl};`,
  };

  method(m: 'PUT' | 'POST') {
    this.#init.method = m;

    return this;
  }

  headers(h: Record<string, string>) {
    this.#init.headers = {
      ...this.#init.headers,
      ...h,
    };

    return this;
  }

  body(b: EndpointPayloads[E]['payload']) {
    this.#init.body = JSON.stringify(b);

    return this;
  }

  /** Adds auth headers */
  withAuth(username: string, token: string) {
    this.#init.headers = {
      ...this.#init.headers,
      'X-Username': username,
      Authorization: `Bearer ${token}`,
    };

    this.#init.credentials = 'same-origin';

    return this;
  }

  async fetch(
    username?: string
  ): Promise<ParsedResponse<R> | ParsedResponse<{ error: string }>> {
    const res = await fetch(`${FetchBuilder.serverUrl}/api${this.#endpoint}`, {
      ...this.#init,
      headers: {
        ...FetchBuilder.defaultHeaders,
        ...this.#init.headers,
      },
    });

    const contentType = res.headers.get('content-type');

    const body =
      res.body !== null && contentType?.includes('application/json')
        ? await res.json()
        : {};

    // TBR: Use secure cookies - https://github.com/tauri-apps/wry/issues/444
    if (body.access_token && username) {
      await tauriInvoke('set_access_token', {
        username,
        accessToken: body.access_token,
      });
    } else if (body.access_token) {
      console.error('Unexpected access_token');
    }

    return {
      status: res.status,
      ok: res.status >= 200 && res.status < 300,
      data: body,
    };
  }
}
