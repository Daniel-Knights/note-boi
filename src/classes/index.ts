export * from './appError';
export * from './debounceQueue';
export * from './encryptor';
export * from './fetchBuilder';
export * from './keyStore';
export * from './storage';
export * from './unsyncedNotesManager';
