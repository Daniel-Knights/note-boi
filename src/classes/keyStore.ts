export class KeyStore {
  static readonly #dbName = 'NoteBoi';
  static readonly #storeName = 'crypto-key';

  static #db?: IDBDatabase;
  static #key?: CryptoKey;

  static async #getDb(): Promise<IDBDatabase | void> {
    if (this.#db) return this.#db;

    const openRequest = window.indexedDB.open(this.#dbName);

    openRequest.onupgradeneeded = () => {
      this.#db = openRequest.result;

      if (!this.#db.objectStoreNames.contains(this.#storeName)) {
        this.#db.createObjectStore(this.#storeName);
      }

      console.log('DB upgraded');
    };

    this.#db = await promisifyRequest(openRequest);

    return this.#db ?? console.error('Failed to open IndexedDB');
  }

  static async reset(): Promise<void> {
    if (this.#db) {
      const deleteRequest = this.#db
        .transaction([this.#storeName], 'readwrite', {
          durability: 'strict',
        })
        .objectStore(this.#storeName)
        .clear();

      await promisifyRequest(deleteRequest);

      this.#db?.close();
    }

    this.#db = undefined;
    this.#key = undefined;

    console.log('Store reset');
  }

  static async storeKey(key: CryptoKey): Promise<void> {
    const db = await this.#getDb();
    if (!db) return;

    const transaction = db.transaction([this.#storeName], 'readwrite', {
      durability: 'strict',
    });

    const store = transaction.objectStore(this.#storeName);
    const request = store.put(key, this.#storeName);

    await promisifyRequest(request);

    this.#key = key;

    console.log('Stored key');
  }

  static async getKey(): Promise<CryptoKey | void> {
    if (this.#key) return this.#key;

    const db = await this.#getDb();
    if (!db) return;

    const transaction = db.transaction([this.#storeName], 'readonly', {
      durability: 'strict',
    });

    const store = transaction.objectStore(this.#storeName);
    const request = store.get(this.#storeName);

    return promisifyRequest<CryptoKey>(request);
  }
}

function promisifyRequest<T>(request: IDBRequest<T>): Promise<T> {
  return new Promise((res, rej) => {
    request.onsuccess = () => {
      res(request.result);
    };

    request.onerror = () => {
      console.error(request.error);

      rej(request.error);
    };

    setTimeout(() => {
      rej(new Error('Request timed out'));
    }, 10000);
  });
}
