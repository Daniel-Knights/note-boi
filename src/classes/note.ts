import type Delta from 'quill-delta';

export class Note {
  readonly uuid: string = crypto.randomUUID();
  timestamp = Date.now();
  content: NoteContent = {
    title: '',
    body: '',
    delta: {
      ops: [],
    },
  };
}

type NoteContent = {
  title: string;
  body: string;
  delta: Partial<Delta>;
};
