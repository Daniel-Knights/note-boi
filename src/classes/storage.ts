import { DeletedNote } from '../api';
import { ColourTheme, UpdateStrategy } from '../constant';

export const STORAGE_KEYS_STRING = {
  MENU_WIDTH: 'note-menu-width',
  THEME: 'theme',
  UPDATE_SEEN: 'update-seen',
  UPDATE_STRATEGY: 'update-strategy',
  USERNAME: 'username',
} satisfies Record<StorageKeyString, string>;

export const STORAGE_KEYS_JSON = {
  UNSYNCED: 'unsynced-note-uuids',
} satisfies Record<StorageKeyJson, string>;

export const STORAGE_KEYS = { ...STORAGE_KEYS_STRING, ...STORAGE_KEYS_JSON };

/** `localStorage` convenience methods with stronger typing. */
export class Storage {
  static get<T extends keyof StorageValuesString>(key: T): StorageValues[T] | null {
    return localStorage.getItem(STORAGE_KEYS[key]) as StorageValues[T] | null;
  }

  static set<T extends keyof StorageValuesString>(key: T, value: StorageValues[T]): void {
    return localStorage.setItem(STORAGE_KEYS[key], value as string);
  }

  static remove(key: StorageKey) {
    return localStorage.removeItem(STORAGE_KEYS[key]);
  }

  static clear() {
    return localStorage.clear();
  }

  static getJson<T extends keyof StorageValuesJson>(key: T): StorageValues[T] | null {
    const storedItem = localStorage.getItem(STORAGE_KEYS[key]);

    return storedItem ? JSON.parse(storedItem) : null;
  }

  static setJson<T extends keyof StorageValuesJson>(key: T, value: StorageValues[T]) {
    return localStorage.setItem(STORAGE_KEYS[key], JSON.stringify(value));
  }
}

//// Types

type StorageValues = {
  MENU_WIDTH: string;
  THEME: ColourTheme;
  UNSYNCED: {
    new: string;
    edited: string[];
    deleted: DeletedNote[];
  };
  UPDATE_SEEN: string;
  UPDATE_STRATEGY: UpdateStrategy;
  USERNAME: string;
};

export type StorageKey = keyof StorageValues;

type StorageValuesString = {
  [K in StorageKey as StorageValues[K] extends string ? K : never]: StorageValues[K];
};

type StorageValuesJson = {
  [K in StorageKey as StorageValues[K] extends object ? K : never]: StorageValues[K];
};

export type StorageKeyString = keyof StorageValuesString;
export type StorageKeyJson = keyof StorageValuesJson;
