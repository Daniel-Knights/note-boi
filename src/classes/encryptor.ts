import { Note } from '../store/note';

import { KeyStore } from './keyStore';

export type EncryptedNote = Omit<Note, 'content'> & {
  content: string;
};

const SALT_LENGTH = 16;
const IV_LENGTH = 12;

const enc = new TextEncoder();
const dec = new TextDecoder();

function buffToBase64(buff: Uint8Array): string {
  return window.btoa(buff.reduce((data, byte) => data + String.fromCharCode(byte), ''));
}

function base64ToBuff(b64: string): Uint8Array {
  return Uint8Array.from(window.atob(b64), (c) => c.charCodeAt(0));
}

export class Encryptor {
  static #generatePasswordKey(password: string) {
    return crypto.subtle.importKey('raw', enc.encode(password), 'PBKDF2', false, [
      'deriveKey',
    ]);
  }

  static async setPasswordKey(password: string): Promise<void> {
    const passwordKey = await this.#generatePasswordKey(password);

    await KeyStore.storeKey(passwordKey);
  }

  static #deriveKey(
    passwordKey: CryptoKey,
    salt: Uint8Array,
    keyUsage: Array<'encrypt' | 'decrypt'>
  ): Promise<CryptoKey> {
    return crypto.subtle.deriveKey(
      {
        name: 'PBKDF2',
        salt,
        iterations: 100000,
        hash: 'SHA-256',
      },
      passwordKey,
      { name: 'AES-GCM', length: 256 },
      false,
      keyUsage
    );
  }

  static async #encryptData(secretData: string, passwordKey: CryptoKey): Promise<string> {
    const salt = crypto.getRandomValues(new Uint8Array(SALT_LENGTH));
    const iv = crypto.getRandomValues(new Uint8Array(IV_LENGTH));
    const derivedKey = await this.#deriveKey(passwordKey, salt, ['encrypt']);
    const encryptedContent = await crypto.subtle.encrypt(
      {
        name: 'AES-GCM',
        iv,
      },
      derivedKey,
      enc.encode(secretData)
    );

    const encryptedContentArr = new Uint8Array(encryptedContent);
    const buff = new Uint8Array(SALT_LENGTH + IV_LENGTH + encryptedContentArr.byteLength);

    buff.set(salt, 0);
    buff.set(iv, SALT_LENGTH);
    buff.set(encryptedContentArr, SALT_LENGTH + IV_LENGTH);

    return buffToBase64(buff);
  }

  static async #decryptData(
    encryptedData: string,
    passwordKey: CryptoKey
  ): Promise<string> {
    const encryptedDataBuff = base64ToBuff(encryptedData);
    const salt = encryptedDataBuff.slice(0, SALT_LENGTH);
    const iv = encryptedDataBuff.slice(SALT_LENGTH, SALT_LENGTH + IV_LENGTH);
    const data = encryptedDataBuff.slice(SALT_LENGTH + IV_LENGTH);
    const derivedKey = await this.#deriveKey(passwordKey, salt, ['decrypt']);
    const decryptedContent = await crypto.subtle.decrypt(
      {
        name: 'AES-GCM',
        iv,
      },
      derivedKey,
      data
    );

    return dec.decode(decryptedContent);
  }

  static async encrypt(plaintext: string, password?: string): Promise<string> {
    if (password) {
      await this.setPasswordKey(password);
    }

    const passwordKey = await KeyStore.getKey();

    return this.#encryptData(plaintext, passwordKey);
  }

  static async decrypt(ciphertext: string, password?: string): Promise<string> {
    if (password) {
      await this.setPasswordKey(password);
    }

    const passwordKey = await KeyStore.getKey();

    return this.#decryptData(ciphertext, passwordKey);
  }

  static async encryptNotes(notes: Note[], password?: string): Promise<EncryptedNote[]> {
    if (password) {
      await this.setPasswordKey(password);
    }

    const passwordKey = await KeyStore.getKey();

    const encryptedNotePromises = notes.map(async (nt) => {
      const encryptedNoteContent = await this.#encryptData(
        JSON.stringify(nt.content),
        passwordKey
      );

      const encryptedNote: EncryptedNote = {
        ...nt,
        content: encryptedNoteContent,
      };

      return encryptedNote;
    });

    return Promise.all(encryptedNotePromises);
  }

  static async decryptNotes(
    notes: (EncryptedNote | Note)[],
    password?: string
  ): Promise<Note[]> {
    if (password) {
      await this.setPasswordKey(password);
    }

    const passwordKey = await KeyStore.getKey();

    const decryptedNotePromises = notes.map(async (nt) => {
      if (typeof nt.content !== 'string') {
        return nt as Note;
      }

      const decryptedNoteContent = await this.#decryptData(nt.content, passwordKey);

      const decryptedNote: Note = {
        ...nt,
        content: JSON.parse(decryptedNoteContent),
      };

      return decryptedNote;
    });

    return Promise.all(decryptedNotePromises);
  }
}
