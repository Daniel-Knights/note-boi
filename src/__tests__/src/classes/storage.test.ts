import { Storage } from '../../../classes';

describe('Storage', () => {
  it('get/set', () => {
    assert.isNull(Storage.get('USERNAME'));

    Storage.set('USERNAME', 'd');

    assert.strictEqual(Storage.get('USERNAME'), 'd');
  });

  it('remove', () => {
    Storage.set('USERNAME', 'd');

    Storage.remove('USERNAME');

    assert.isNull(Storage.get('USERNAME'));
  });

  it('clear', () => {
    Storage.set('USERNAME', 'd');
    Storage.set('THEME', 'Dark');

    Storage.clear();

    assert.isNull(Storage.get('USERNAME'));
    assert.isNull(Storage.get('THEME'));
  });

  it('setJson/getJson', () => {
    assert.isNull(Storage.getJson('UNSYNCED'));

    const unsynced = {
      new: 'note1',
      edited: ['note2'],
      deleted: [{ id: 'note3', deleted_at: 0 }],
    };

    Storage.setJson('UNSYNCED', unsynced);

    assert.deepEqual(Storage.getJson('UNSYNCED'), unsynced);
  });
});
