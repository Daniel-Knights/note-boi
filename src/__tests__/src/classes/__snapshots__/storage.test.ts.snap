// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Storage > clear > KeyStore.getKey() 1`] = `undefined`;

exports[`Storage > clear > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Storage > clear > Storage.get('THEME') 1`] = `null`;

exports[`Storage > clear > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Storage > clear > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Storage > clear > Storage.get('USERNAME') 1`] = `null`;

exports[`Storage > clear > allCalls 1`] = `[]`;

exports[`Storage > clear > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Storage > clear > mockDb 1`] = `
{
  "deletedNoteIds": Set {},
  "encryptedNotes": [
    {
      "content": "content",
      "id": "2cbe0086-4b6f-466f-9ec0-9a302480da6a",
      "timestamp": 958201892455,
    },
    {
      "content": "content",
      "id": "2e36d871-0de2-43e6-9cbc-3c77e5d154cb",
      "timestamp": 811831863971,
    },
    {
      "content": "content",
      "id": "b8ff4ac5-6541-49d2-9c79-4444d872481e",
      "timestamp": 739056902017,
    },
    {
      "content": "content",
      "id": "6b24cde1-6118-4277-9791-8eab3c71cabb",
      "timestamp": 157836992432,
    },
    {
      "content": "content",
      "id": "ac4e006d-38d4-4373-af31-4a5f46fc5184",
      "timestamp": 1066734006578,
    },
    {
      "content": "content",
      "id": "6fb3ce26-82e0-408f-84bf-bd8cd64c35b2",
      "timestamp": 1642640862249,
    },
    {
      "content": "content",
      "id": "81648517-86ff-4ed9-8bf7-941c32dd0fad",
      "timestamp": 242282290063,
    },
    {
      "content": "content",
      "id": "9a2f52b3-69ac-4158-a9f7-b79ba9dbafd3",
      "timestamp": 550235579146,
    },
    {
      "content": "content",
      "id": "b9754a63-843c-4f59-9aea-9a374aa1603d",
      "timestamp": 1533441892487,
    },
    {
      "content": "content",
      "id": "7a7bccda-3fac-4465-b9c9-8f67dc1b5bdf",
      "timestamp": 1650008403223,
    },
  ],
  "users": {
    "d": "1",
  },
}
`;

exports[`Storage > clear > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Storage > clear > openedPopup 1`] = `undefined`;

exports[`Storage > clear > selectedTheme 1`] = `"System"`;

exports[`Storage > clear > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "set": [Function],
    "size": 0,
    "store": [Function],
  },
  "username": "",
}
`;

exports[`Storage > clear > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Storage > get/set > KeyStore.getKey() 1`] = `undefined`;

exports[`Storage > get/set > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Storage > get/set > Storage.get('THEME') 1`] = `null`;

exports[`Storage > get/set > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Storage > get/set > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Storage > get/set > Storage.get('USERNAME') 1`] = `"d"`;

exports[`Storage > get/set > allCalls 1`] = `[]`;

exports[`Storage > get/set > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Storage > get/set > mockDb 1`] = `
{
  "deletedNoteIds": Set {},
  "encryptedNotes": [
    {
      "content": "content",
      "id": "2cbe0086-4b6f-466f-9ec0-9a302480da6a",
      "timestamp": 958201892455,
    },
    {
      "content": "content",
      "id": "2e36d871-0de2-43e6-9cbc-3c77e5d154cb",
      "timestamp": 811831863971,
    },
    {
      "content": "content",
      "id": "b8ff4ac5-6541-49d2-9c79-4444d872481e",
      "timestamp": 739056902017,
    },
    {
      "content": "content",
      "id": "6b24cde1-6118-4277-9791-8eab3c71cabb",
      "timestamp": 157836992432,
    },
    {
      "content": "content",
      "id": "ac4e006d-38d4-4373-af31-4a5f46fc5184",
      "timestamp": 1066734006578,
    },
    {
      "content": "content",
      "id": "6fb3ce26-82e0-408f-84bf-bd8cd64c35b2",
      "timestamp": 1642640862249,
    },
    {
      "content": "content",
      "id": "81648517-86ff-4ed9-8bf7-941c32dd0fad",
      "timestamp": 242282290063,
    },
    {
      "content": "content",
      "id": "9a2f52b3-69ac-4158-a9f7-b79ba9dbafd3",
      "timestamp": 550235579146,
    },
    {
      "content": "content",
      "id": "b9754a63-843c-4f59-9aea-9a374aa1603d",
      "timestamp": 1533441892487,
    },
    {
      "content": "content",
      "id": "7a7bccda-3fac-4465-b9c9-8f67dc1b5bdf",
      "timestamp": 1650008403223,
    },
  ],
  "users": {
    "d": "1",
  },
}
`;

exports[`Storage > get/set > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Storage > get/set > openedPopup 1`] = `undefined`;

exports[`Storage > get/set > selectedTheme 1`] = `"System"`;

exports[`Storage > get/set > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "set": [Function],
    "size": 0,
    "store": [Function],
  },
  "username": "",
}
`;

exports[`Storage > get/set > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Storage > remove > KeyStore.getKey() 1`] = `undefined`;

exports[`Storage > remove > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Storage > remove > Storage.get('THEME') 1`] = `null`;

exports[`Storage > remove > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Storage > remove > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Storage > remove > Storage.get('USERNAME') 1`] = `null`;

exports[`Storage > remove > allCalls 1`] = `[]`;

exports[`Storage > remove > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Storage > remove > mockDb 1`] = `
{
  "deletedNoteIds": Set {},
  "encryptedNotes": [
    {
      "content": "content",
      "id": "2cbe0086-4b6f-466f-9ec0-9a302480da6a",
      "timestamp": 958201892455,
    },
    {
      "content": "content",
      "id": "2e36d871-0de2-43e6-9cbc-3c77e5d154cb",
      "timestamp": 811831863971,
    },
    {
      "content": "content",
      "id": "b8ff4ac5-6541-49d2-9c79-4444d872481e",
      "timestamp": 739056902017,
    },
    {
      "content": "content",
      "id": "6b24cde1-6118-4277-9791-8eab3c71cabb",
      "timestamp": 157836992432,
    },
    {
      "content": "content",
      "id": "ac4e006d-38d4-4373-af31-4a5f46fc5184",
      "timestamp": 1066734006578,
    },
    {
      "content": "content",
      "id": "6fb3ce26-82e0-408f-84bf-bd8cd64c35b2",
      "timestamp": 1642640862249,
    },
    {
      "content": "content",
      "id": "81648517-86ff-4ed9-8bf7-941c32dd0fad",
      "timestamp": 242282290063,
    },
    {
      "content": "content",
      "id": "9a2f52b3-69ac-4158-a9f7-b79ba9dbafd3",
      "timestamp": 550235579146,
    },
    {
      "content": "content",
      "id": "b9754a63-843c-4f59-9aea-9a374aa1603d",
      "timestamp": 1533441892487,
    },
    {
      "content": "content",
      "id": "7a7bccda-3fac-4465-b9c9-8f67dc1b5bdf",
      "timestamp": 1650008403223,
    },
  ],
  "users": {
    "d": "1",
  },
}
`;

exports[`Storage > remove > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Storage > remove > openedPopup 1`] = `undefined`;

exports[`Storage > remove > selectedTheme 1`] = `"System"`;

exports[`Storage > remove > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "set": [Function],
    "size": 0,
    "store": [Function],
  },
  "username": "",
}
`;

exports[`Storage > remove > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Storage > setJson/getJson > KeyStore.getKey() 1`] = `undefined`;

exports[`Storage > setJson/getJson > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Storage > setJson/getJson > Storage.get('THEME') 1`] = `null`;

exports[`Storage > setJson/getJson > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Storage > setJson/getJson > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Storage > setJson/getJson > Storage.get('USERNAME') 1`] = `null`;

exports[`Storage > setJson/getJson > allCalls 1`] = `[]`;

exports[`Storage > setJson/getJson > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Storage > setJson/getJson > mockDb 1`] = `
{
  "deletedNoteIds": Set {},
  "encryptedNotes": [
    {
      "content": "content",
      "id": "2cbe0086-4b6f-466f-9ec0-9a302480da6a",
      "timestamp": 958201892455,
    },
    {
      "content": "content",
      "id": "2e36d871-0de2-43e6-9cbc-3c77e5d154cb",
      "timestamp": 811831863971,
    },
    {
      "content": "content",
      "id": "b8ff4ac5-6541-49d2-9c79-4444d872481e",
      "timestamp": 739056902017,
    },
    {
      "content": "content",
      "id": "6b24cde1-6118-4277-9791-8eab3c71cabb",
      "timestamp": 157836992432,
    },
    {
      "content": "content",
      "id": "ac4e006d-38d4-4373-af31-4a5f46fc5184",
      "timestamp": 1066734006578,
    },
    {
      "content": "content",
      "id": "6fb3ce26-82e0-408f-84bf-bd8cd64c35b2",
      "timestamp": 1642640862249,
    },
    {
      "content": "content",
      "id": "81648517-86ff-4ed9-8bf7-941c32dd0fad",
      "timestamp": 242282290063,
    },
    {
      "content": "content",
      "id": "9a2f52b3-69ac-4158-a9f7-b79ba9dbafd3",
      "timestamp": 550235579146,
    },
    {
      "content": "content",
      "id": "b9754a63-843c-4f59-9aea-9a374aa1603d",
      "timestamp": 1533441892487,
    },
    {
      "content": "content",
      "id": "7a7bccda-3fac-4465-b9c9-8f67dc1b5bdf",
      "timestamp": 1650008403223,
    },
  ],
  "users": {
    "d": "1",
  },
}
`;

exports[`Storage > setJson/getJson > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Storage > setJson/getJson > openedPopup 1`] = `undefined`;

exports[`Storage > setJson/getJson > selectedTheme 1`] = `"System"`;

exports[`Storage > setJson/getJson > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "set": [Function],
    "size": 0,
    "store": [Function],
  },
  "username": "",
}
`;

exports[`Storage > setJson/getJson > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;
